File: src/lib/scrapers/hn-hiring.ts
import type { RawPosting } from "@/types";
import { BaseScraper } from "./base";

const MEDICAL_TITLE_KEYWORDS = [
  "medical receptionist", "front desk", "patient service", "patient access",
  "appointment scheduler", "scheduling coordinator", "patient coordinator",
  "patient care coordinator", "patient intake", "intake coordinator",
  "medical administrative", "medical office assistant", "medical secretary",
  "medical records", "health information", "data entry clerk",
  "insurance verification", "eligibility", "prior authorization",
  "medical biller", "medical billing", "accounts receivable",
  "claims processor", "revenue cycle", "collections specialist",
  "referral coordinator", "dental receptionist", "patient recall",
];

interface HnSearchHit {
  objectID: string;
  title?: string;
  created_at_i?: number;
}

interface HnComment {
  id: number;
  text?: string;
  author?: string;
  created_at_i?: number;
  children?: HnComment[];
}

export class HnHiringScraper extends BaseScraper {
  constructor(contactEmail: string) {
    super("hn", contactEmail);
  }

  async fetch(limit: number): Promise<RawPosting[]> {
    const searchUrl =
      "https://hn.algolia.com/api/v1/search?query=who+is+hiring&tags=story";
    const searchRes = await this.fetchWithRetry(searchUrl);
    const searchData = (await searchRes.json()) as {
      hits: HnSearchHit[];
    };
    const latest = searchData.hits[0];
    if (!latest) return [];

    await this.paginatedJitter();
    const threadRes = await this.fetchWithRetry(
      `https://hn.algolia.com/api/v1/items/${latest.objectID}`
    );
    const thread = (await threadRes.json()) as { children?: HnComment[] };

    const postings: RawPosting[] = [];
    const walk = (comments: HnComment[] | undefined) => {
      if (!comments) return;
      for (const comment of comments) {
        const text = comment.text ?? "";
        if (text.length < 40) {
          walk(comment.children);
          continue;
        }
        const titleMatch = text.match(/(?:^|\|)\s*([^|]+?)\s*\|/);
        const title = titleMatch?.[1]?.trim() ?? `HN comment ${comment.id}`;
        postings.push({
          source: "hn",
          externalId: String(comment.id),
          url: `https://news.ycombinator.com/item?id=${comment.id}`,
          title,
          company: comment.author ?? "unknown",
          location: "Remote",
          description: text.replace(/<[^>]+>/g, " "),
          postedAt: comment.created_at_i
            ? new Date(comment.created_at_i * 1000)
            : null,
          raw: { id: comment.id, author: comment.author, text },
        });
        walk(comment.children);
      }
    };

    walk(thread.children);
    const matched = postings.filter((p) => {
      const lower = (p.title + " " + p.description).toLowerCase();
      return MEDICAL_TITLE_KEYWORDS.some((kw) => lower.includes(kw));
    });
    return matched.slice(0, limit);
  }
}

File: scraper/scrapers/hn_hiring.py
import re
from typing import Any

from .base import BaseScraper


class HnHiringScraper(BaseScraper):
    def __init__(self, contact_email: str):
        super().__init__("hn", contact_email)

    def fetch(self, limit: int) -> list[dict[str, Any]]:
        search_url = (
            "https://hn.algolia.com/api/v1/search?query=who+is+hiring&tags=story"
        )
        search_res = self._fetch_with_retry(search_url)
        hits = search_res.json().get("hits") or []
        if not hits:
            return []

        latest_id = hits[0].get("objectID")
        self._jitter()
        thread_res = self._fetch_with_retry(
            f"https://hn.algolia.com/api/v1/items/{latest_id}"
        )
        thread = thread_res.json()

        postings: list[dict[str, Any]] = []

        def walk(comments: list[dict[str, Any]] | None) -> None:
            if not comments or len(postings) >= limit:
                return
            for comment in comments:
                if len(postings) >= limit:
                    return
                text = comment.get("text") or ""
                if len(text) < 40:
                    walk(comment.get("children"))
                    continue

                title_match = re.search(r"(?:^|\|)\s*([^|]+?)\s*\|", text)
                title = (
                    title_match.group(1).strip()
                    if title_match
                    else f"HN comment {comment.get('id')}"
                )
                clean = re.sub(r"<[^>]+>", " ", text)
                postings.append(
                    {
                        "source": "hn",
                        "externalId": str(comment.get("id")),
                        "url": f"https://news.ycombinator.com/item?id={comment.get('id')}",
                        "title": title,
                        "company": comment.get("author") or "unknown",
                        "location": "Remote",
                        "description": clean,
                        "postedAt": comment.get("created_at_i"),
                        "raw": {
                            "id": comment.get("id"),
                            "author": comment.get("author"),
                            "text": text,
                        },
                    }
                )
                walk(comment.get("children"))

        walk(thread.get("children"))
        return postings[:limit]
